### ISSUE TITLE
Core/Spells: Spell::SearchTargets doesn't always return full list of targets

### ISSUE THREAD

**Issue by Zhalfyr** [TrinityCore/TrinityCore#22375]
Description: 
it appears that  if a circle with radius equals to the range of the casted spell is entirely in your cell the visitors don't search for targets in adjacent cells.
Let me take an example : if you're standing 10 meters away from the nearest border of your cell, and your target is standing a bit further, behind that border. if the creature as an object size that make him overlap in your cell enough to be in range of an AoE you would cast, you won't hit it unless the area of the spell overlap the creature's cell, although you should hit the creature since it's objectsize makes it in range.

Steps to reproduce the problem:
Spawn a creature with huge radius (more than cellsize to be sure you're not in its cell)  go the furthest you can where your wow client says you're still in melee range (you can cast melee targeted attacks on the creature) and try to cast an AoE centered on you, you won't hit the creature.

Branch(es): master

TC rev. hash/commit: rev. 5a0ce66

OS : W10 x64

**Comment by wonder2011** [comment]
Omg, that's bad !

**Comment by matanshukry** [comment]
A bit old, but anyone got an actual example of such a creature with a "huge radius" ?

**Comment by Zhalfyr** [comment]
Try Fallen Avatar in Tomb of Sargeras. If he is at a grid intersection, you can experience the issue.

**Comment by matanshukry** [comment]
@Zhalfyr any way to make sure / see / test if I'm in a grid intersection?

**Comment by Zhalfyr** [comment]
Its original location is at a grid intersection. 
Entry : 116939
X : 6606.67
Y : -795.122
Z : 1665.61
O : 3.13905
MapId : 1676

**Comment by matanshukry** [comment]
@Zhalfyr Not sure how to get in?

`.go xyz 6606.67 -795.122 1665.61 1676` results in `invalid target map or coordinates..`
Tried to walk in too, but looks like areatrigger isn't implemented.

Got perhaps another example, something that can be reproduced easily?

**Comment by Shauren** [comment]
insert a dummy instance_template row

**Comment by matanshukry** [comment]
Still problematic to check though:
- The position is in the air. Probably should have a gameobject, doesn't exist there
- the npc it self is friendly so can't attack it either

**Comment by matanshukry** [comment]
Closest thing I sort of did:
1. `.cheat god`
2. `.go xyz 6606.67 -795.122 1665.61 1676`
3. `.npc add temp 36853` ( Sindragosa )
4. Quickly attack it, hopefully will be as close position as possible
5. Used priest `Holy Nova`. Does aoe damage around the caster - Sindragosa is hit

**Comment by Zhalfyr** [comment]
1. `.gm on`
2. `.go 6604.237-805.618 1509.95 1676`
3. `.npc add 50203 (Al'Akir)`
4. `.go 6576.613 -774.980 1509.91`
Stay in GM on to avoid it moving.
you should be able to melee attack it but your Holy nova won't hit.

**Comment by matanshukry** [comment]
@Zhalfyr to clarify, I used these commands:

1. `.go xyz 6604.237 -805.618 1509.95 1676`
3. `.go xyz 6576.613 -774.980 1509.91`

But assuming that's correct, the range is bad.
I used `.dist` command at the last position, and distance was 41 yards. Holy nova max range is 12 yards.
I got close to 11 yards and it then hit it.

p.s.
Al'Akir isn't friendly so it didn't actually hit him, but I put a breakpoint on selecting targets and I noticed it did check for him when I was close, and didn't when I was far.

**Comment by Zhalfyr** [comment]
Range of the spell take in account the radius of the target.
That's why it's usefull to compare with a melee swing since it has a 5y range. If the client allow you to melee, you should always hit with Holy Nova.
You can try spawning the target at the pos `.go 6597.56 -798.15 1509.91` I'm pretty sure from the same location than before ( `.go 6576.613 -774.980 1509.91`) your Holy Nova will correctly hit. because these 2 positions are in the same Cell

**Comment by matanshukry** [comment]
> Range of the spell take in account the radius of the target.
No it doesn't; but that could be the issue. Looking into it.

**Comment by Zhalfyr** [comment]
If it doesn't, you can never hit large bosses with AoE if you can't go under them.
Bosses like Fallen avatar or Krosus in Nighthold for example.

**Comment by Zhalfyr** [comment]
Example also works for Kologarn I think if you want TLK stuff 😄

**Comment by matanshukry** [comment]
@Zhalfyr When doing melee attack, you check for range between 2 units, so you can easily check the range of each one - which we do.

When you want to check for units around you, you can't check the target radius because you don't have any targets yet - that is exactly what you're looking for.

Also know the size of the creature isn't big, but only it's reach size (at least for Al'Akir).
That is:
- BoundingRadius: 0.699999988
- CombatReach: 40.0000000

But yeah I believe it needs fixin;g; I'm on it.

**Comment by Zhalfyr** [comment]
Actually, you can enable using the size of object when checking for targets iirc.
The searcher looks for all the entities in all the cells which your searching radius overlap.
With this size checking option, you can find Alakir for exemple even outside the 12y, given that he is in a cell withing your searching radius.

Now from what I checked when I opened the issue (so we're talking 2years ago memories 😃) the alternative I was studying was register units not only in their own cells, but aswell in the cells their bounding radius overlap with. This way would allow the searcher to still search the same cells as now, but taking in consideration some creatures from neighbour cells (those who overlap).
If you got other idea of fix or a good way of implementing the one I offer, I'm interested.

**Comment by matanshukry** [comment]
Actually the 2nd idea you had (putting npcs on multiple cells) is exactly the idea that came into my mind.
But I still need to look into it more.

**Comment by Zhalfyr** [comment]
Alright, if you have any question or want to discuss an idea I'll be around on the discord.
I studied the system and the issue pretty extensively at the time.

**Comment by offl** [comment]
Is is this about wrongly calculated combat reach?
'Player combat reach + creature combat reach = attack distance'  instead of 'creature combat reach = attack distance'

**Comment by Zhalfyr** [comment]
Not really, we're talking about AoE spell targeting which doesn't use  attack distance

**Comment by matanshukry** [comment]
@offl nope. Combat reach is calculated fine. The problem is when you look for units that you don't know their combat reach.

Let's say you look for units in distance 10 yards - which is the spell radius.
Then you got a creature that's 20 yards away, but the creature got reach of 30 yards.

You should reach him, but the search functionality won't find him since he's too far away.

I made it simple, we're using grids rather than just locations - but the idea remains.
